swarm.js

```diff
@@ -1,38 +1,146 @@
 'use strict'
 const peerInfo = require('./lib/peer-info')
+const peerQueue = require('./lib/queue')
 const { EventEmitter } = require('events')
 const guts = require('@hyperswarm/guts')
 
+const MAX_PEERS_DEFAULT = 16
+const ERR_MISSING_KEY = 'key is required and must be a buffer'
+
 module.exports = opts => new Swarm(opts)
 
 class Swarm extends EventEmitter {
   constructor (opts = {}) {
     super()
-    const { bootstrap, ephemeral } = opts
-    this.network = guts({
+    const {
+      maxPeers = MAX_PEERS_DEFAULT,
+      bootstrap,
+      ephemeral
+    } = opts
+    const queue = peerQueue()
+    const network = guts({
       bootstrap,
       ephemeral,
-      bind: () => {
-        this.emit('listening')
-      },
+      bind: () => this.emit('listening'),
       socket: (socket, isTCP) => {
         const info = peerInfo(null)
         info.connected(socket, isTCP)
         this.emit('connection', socket, info)
       },
-      close: () => {
-        this.emit('close')
-      }
+      close: () => this.emit('close')
     })
+    queue.on('readable', this._drain(queue))
+    this.peers = 0
+    this.maxPeers = maxPeers
+    this.emphemeral = ephemeral !== false
+    this.network = network
+    this.queue = queue
+  }
+  _drain (queue) {
+    const onConnect = (info) => (err, socket, isTCP) => {
+      if (err) {
+        this.peers -= 1
+        if (!queue.requeue(info)) queue.requeue(info)
```

Comment:
what's happening on this line?

Reply:
we're sending it back round the queue in case the error is non-permanent - eventually it'll work itself out via depriotization -> ban right?

Reply:
ohhh wait.. the conditional check is wrong, it's meant to be checking something else I need to remember what it is supposed to be checking. 

Is it correct to requeue at this point thought?

Reply:
Right to requeue for sure. I actually think you just want to always requeue here.

Reply:
yeah cool ok - whatever I meant to be checking (maybe reconnecting?) I think it should be unnecessary anyway, the queue itself should take care of any conditional logic